flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java

```diff
@@ -792,12 +832,40 @@ else if (t instanceof Class) {
 
 		return null;
 	}
-	
+
+	@SuppressWarnings({"unchecked", "rawtypes"})
 	private <IN1> TypeInformation<?> createTypeInfoFromInput(TypeVariable<?> returnTypeVar, ArrayList<Type> inputTypeHierarchy, Type inType, TypeInformation<IN1> inTypeInfo) {
 		TypeInformation<?> info = null;
-		
+
+		// use a factory to find corresponding type information to type variable
+		final ArrayList<Type> factoryHierarchy = new ArrayList<>(inputTypeHierarchy);
+		final TypeInfoFactory<?> factory = getClosestFactory(factoryHierarchy, inType);
+		if (factory != null) {
+			// the type that defines the factory is last in factory hierarchy
+			final Type factoryDefiningType = factoryHierarchy.get(factoryHierarchy.size() - 1);
+			// defining type has generics, the factory need to be asked for a mapping of subtypes to type information
+			if (factoryDefiningType instanceof ParameterizedType) {
```

Comment:
Yes, because we try to create type information from the input in this method. This only works if a type variable connects input with output. `Map<InType<A>, OutType<A>>`